### DISCUSSION TITLE
withClientName internal implementation changes from 5.x to 6.x

### DISCUSSION THREAD

**Discussion by Akaame** [redis/lettuce#1966] [Q&A] [answered]
We have upgraded a dependency that itself depended on `lettuce-core` so an upgrade from `5.1.6` to `6.1.1` versions has taken place.

We are creating the Redis URI to the "managed GCP Redis" as follows: 

```java
private RedisClient redisClient() {
  return RedisClient.create(RedisURI.builder()
    .withClientName(properties.getRedis().getClientName())
    .withHost(properties.getRedis().getHost())
    .withPort(properties.getRedis().getPort())
    .build());
}
```

Earlier the `withClientName` option was working as expected. And with the new version it issues a `CLIENT SETNAME foo` command which is prohibited by GCP's Redis. Therefore, Redis connection setup fails.

<img width="523" alt="image" src="https://user-images.githubusercontent.com/11539091/149139579-4d900e21-d526-4c2d-b3eb-0144b3ee2d41.png">

Questions:

1. How was this working earlier?

Some hints I have found the understand this change in behavior.

```java
/**
* @param clientName
* @deprecated since 6.0, use {@link RedisAsyncCommands#clientSetname(Object)}.
*/
@Deprecated
public void setClientName(String clientName) {
  CommandArgs<String, String> args = new CommandArgs<>(StringCodec.UTF8).add(CommandKeyword.SETNAME).addValue(clientName);
  AsyncCommand<String, String, String> async = new AsyncCommand<>(
          new Command<>(CommandType.CLIENT, new StatusOutput<>(StringCodec.UTF8), args));
  state.setClientName(clientName);

  dispatch((RedisCommand) async);
}
```

And this PR:
https://github.com/lettuce-io/lettuce-core/commit/abbfdf69f1a71832fa2d4d90a2e010a876f3a9a6

My understanding:
Earlier this was being passed along during connection setup. And now the connection is set up and necessary `CLIENT` subcommands are issued to further customize it. Can someone from the contributors/maintainers verify this?

2. How do we get around this given the limitation from GCP Redis

Naive way would be prefixing every key with `foo:` as I think CLIENT SETNAME does exactly that anyway.

Thanks in advance and cheers.

**Comment by mp911de** [comment, marked as answer]
It looks as if your Redis server rejects the `CLIENT SETNAME` command. If you're using Redis 6, then you could enable RESP3 as RESP3 uses the `HELLO` command to initialize the connection and that takes the client name as well.

**Reply by Akaame** [reply]
Hallo and thanks for the quick answer. If I connect to a Redis 6.x instance with RESP3 enabled, will the client automatically prefer the new protocol or is there a toggle on the client side?

**Reply by mp911de** [reply]
By default, it will try to use RESP3 and fall back to RESP2 if the server doesn't support it. You can force a protocol version through `ClientOptions` and set it via `RedisClient.setOptions(…)`
